doc/contributor/code-patterns.md

```diff
@@ -0,0 +1,52 @@
+# Code Patterns
+
+## Where to allocate helper nodes (a node used in another node)
+
+* If the node does not use the DSL, you should allocate the helper node [lazily](#the-lazy-pattern).
+* If the helper node is used by every specialization: either allocate the helper node eagerly as a `@Child` or use the lazy pattern.
```

Comment:
Is there some guidance on where eager or lazy is better in this case?

Reply:
Not in this document.
I prefer eager when I know the helper node will be always used by the parent, and lazy otherwise.
But always lazy might be better for memory consumption (it's a lot more verbose unfortunately).

Reply:
Not sure what others think. @chrisseaton @nirvdrum ?

Reply:
I agree with @eregon. But it can be tricky to work out. You could have a specialization that always executes a node, which would mean allocating it eagerly here, but the parent may be called from Ruby in a branch that's never taken. In that situation, allocating lazily would have paid off.

I think using the DSL helps out here because you wouldn't allocate the node for a specialization unless it's entered. But that gets messy when all specializations for a method need the same node.

For the time being, I'm okay going with @eregon's guideline. At some point we may make a more concerted effort to reduce memory consumption and we can revisit. But we could probably do that on a case-by-case basis based on split count.